DAP Dagster Refresh Pipeline

quazi-h opened the pull request:
## Why

[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1712)

## This PR
Added a simple dagster refresh pipeline.
There are 3 modes available: local, dev, and prod. The latter 2 use the dataflow runner.
Added tests for each solid, need to add an e2e test (currently using a scratch script to test running the whole pipeline).
There will be some future work to clean up abstractions and replace some of the components here with what we have in dagster_utils.

quazi-h committed:
Added a dap_orchestration folder within orchestration for dagster components.
Set up a poetry project for the new dagster project and used the monster docs to set up the project.
Added a solid to extract records as well as a beam runner resource (copied from HCA - will factor out in the next iteration).

quazi-h committed:
Updated LocalBeamRunner config variable "project" to "scala_project".
Added a transform_records solid.
Add a pipeline to the Dagster project to call the two solids.

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1712-dagster-refresh-pipeline

quazi-h committed:
uploading most recent changes - testing

quazi-h committed:
Added a gcsfs dependency to work with files on gcs.
Removed additional args from the local beam runner resource config and replaced executable code.
Added a solid to write outfiles using the existing python script in the repo.

quazi-h committed:
Added configured solids for hles, cslb, and env based on the base extract_records solid.
Added a new tests script to test each of the configured solids (extract, transform, write_outfiles for hles, cslb, and environment).
Moved scala_project and target_class out of the local_beam_runner resource config
target_class and scala_project will be hardcoded to each configured solid
target_class is unique to each distinct pipeline while scala_project is unique to the ingest steps (extract, transform)

aherbst-broad committed:
Trying something different (#162)

Builder method that returns a SolidDefintion

quazi-h committed:
Refactored the transform solids and added a config builder to handle each pipeline, similar to extract.
Added unit tests for transform solids.

quazi-h committed:
Added a step to the outfiles solid that makes a TSV directory.
We may want to refactor the TSV script so we can import a function and call it on unique pipelines.

quazi-h committed:
Importing the configured transform solids to pipelines.
Added type annotations where needed as indicated by mypy.
Added some todo notes in the test code to indicate issues noted by mypy.

aherbst-broad commented on .github/workflows/validate-python-subdirectory.yaml:
```diff
@@ -0,0 +1,51 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python.
+# Use this workflow if your Python code is in a subdirectory of a non-Python repository.
```
⛏️ this was presumably copied from our python template. I think the comment can be updated or removed as no one will actually use this workflow elsewhere now that it's in DAP.

aherbst-broad commented on orchestration/dap_orchestration/config/__init__.py:
```diff
@@ -0,0 +1,19 @@
+    Helper function for preconfiguring resources, specifically.
+    Tells preconfigure_for_mode to look in `resources/[resource name]` for config files.
+    """
+    return preconfigure_for_mode(resource, mode_name, additional_schema, f'resources/{resource.__name__}')
```
⛏️ newline

aherbst-broad commented on orchestration/dap_orchestration/config/__init__.py:
```diff
@@ -0,0 +1,19 @@
+preconfigure_for_mode = configurator_aimed_at_directory(os.path.dirname(__file__))
+
+
+def preconfigure_resource_for_mode(
```
⛏️ This feels like it should be extracted to `dagster-utils` if it doesn't already exists there. Can you ticket if that's the case?

aherbst-broad commented on orchestration/dap_orchestration/config/resources/local_beam_runner/global.yaml:
```diff
@@ -0,0 +1,8 @@
+worker_machine_type: n1-standard-1
+autoscaling_algorithm: THROUGHPUT_BASED
+num_workers: 4
+max_num_workers: 8
```
⛏️ newline

aherbst-broad commented on orchestration/dap_orchestration/config/resources/local_beam_runner/dev.yaml:
```diff
@@ -0,0 +1 @@
+google_project: broad-dsp-monster-dap-dev
```
⛏️ newline

aherbst-broad commented on orchestration/dap_orchestration/config/resources/local_beam_runner/prod.yaml:
```diff
@@ -0,0 +1 @@
+google_project: broad-dsp-monster-dap-prod
```
⛏️ newline

aherbst-broad commented on orchestration/dap_orchestration/pipelines.py:
```diff
@@ -0,0 +1,34 @@
+    hles_transform_records(hles_extract_records())
+    cslb_transform_records(cslb_extract_records())
+    transform_files=env_transform_records(env_extract_records())
+    write_outfiles(transform_files)
```
🔧 Why does `write_outfiles` only consume the output of `env_transform_records`? It seems like it'd need the output from `hles_` and `cslb_`.

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+    max_num_workers: int
+    google_project: str
+
+    # note: Use "-> None" if function does not return a value
```
⛏️ This comment seems unneeded.

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+    def __post_init__(self) -> None:
+        self.arg_dict = {
+            # resource config - how stuff behaves
+            # will not affect what is returned - will only affect run performance/success
```
⛏️ I think the rest of the comments in this dict are not needed, but I'm 51/49 on that. Feel free to push back.

raaidbroad replied:
+1

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+            "experiments": "shuffle_mode=service",
+        }
+
+    # pull the target_class out of the resource config
```
⛏️ This comment is unmoored -- what does it refer to? If it's outdated, remove.

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+        # create a new dictionary containing the keys and values of arg_dict + solid arguments
+        dataflow_run_flags = {**self.arg_dict, **run_arg_dict}
+        self.logger.info("Dataflow beam runner")
+        # pull target_class out of list of flags
```
⛏️ I think this comment as-is is of similarly little value; I _do_ feel one is warranted for these special cases though. Perhaps:

```
# special case target_class and scala_project as those cannot be placed in the args list of the sbt invocation`?

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+    working_dir: str
+    logger: DagsterLogManager
+
+    # todo: **kwargs param to local beam runner + dataflow beam runner when we refactor this to use dagster_utils
```
⛏️ Remove this comment if it's not serving a purpose

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+            "runner": "direct",
+        }
+
+    # pull the target_class out of the resource config
```
⛏️ Remove

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+            run_arg_dict: dict[str, Any],
+    ) -> None:
+        # create a new dictionary containing the keys and values of arg_dict + solid arguments
+        dataflow_run_flags = {**self.arg_dict, **run_arg_dict}
```
🤔 This and the dataflow beam runner are almost the same in implementation, it appears they differ solely in the flags they consume. I'm wondering if we can extract a common class. Probably out of scope for this PR, but worth a small bite ticket so we don't forget.

aherbst-broad replied:
🔧 This var is named confusingly, this isn't actually a dataflow runner.

aherbst-broad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,135 @@
+
+class TestBeamRunner:
+    def run(self, arg_dict: dict[str, Any]) -> None:
+        # no thoughts, head empty
```
⛏️ Remove this comment

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,122 @@
+from dagster.core.execution.context.compute import AbstractComputeExecutionContext
+
+
+## A solid is a unit of computation that yields a stream of events (similar to a function), while a composite is a collection of solids.
```
⛏️ Same note about removing comments like this

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,122 @@
+        "scala_project": "dog-aging-hles-extraction"
+    }
+    context.resources.beam_runner.run(arg_dict)
+    # todo: resolve mypy error: Returning Any from function declared to return "str"
```
⛏️ Remove when this error is resolved.

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,122 @@
+        "scala_project": "dog-aging-hles-transformation"
+    }
+    context.resources.beam_runner.run(arg_dict)
+    # todo: resolve mypy error: Returning Any from function declared to return "str"
```
⛏️ Remove when this error is resolved.

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,122 @@
+        cwd=context.solid_config["working_dir"]
+    )
+    # todo: resolve mypy error: Returning Any from function declared to return "str"
+    return context.solid_config["working_dir"]+"/tsv_output"
```
⛏️ Prefer f-strings, like so:
```suggestion
    return f"{context.solid_config["working_dir"]}/tsv_output"
```

aherbst-broad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,122 @@
+    }
+
+
+# todo: how do I specify the input_values (output_prefix) for a configured function?
```
🤔 Is this still an issue? I believe it should be passed in via the config at runtime, but I'm not 100% on that. Happy to pair up and figure this out.

aherbst-broad reviewed (commented):
Looking generally good, I think there's some cleanup to be done before merging. Let me know if you have any questions.

quazi-h committed:
Added newlines where needed, cleaned up comments
Renamed a variable within local beam runner to correctly describe contents
Replaced some statements with f strings

quazi-h committed:
Added "input_prefix" for solids instead of returning strings (returning None)

quazi-h committed:
Updated solids to take in a working dir and each configured solid (hles_extract, cslb_, env_) will specify the subdirectory within the working dir to create and use.
Updated solids tests to remove personal directories.
Added a resource to return the refresh_directory.
Added a resource for outfile writer and moved code to create outfiles into that resource rather than running a subprocess call within a solid.
Updates made to address mypy errors.

raaidbroad commented on orchestration/pyproject.toml:
```diff
@@ -0,0 +1,27 @@
+name = "dap_orchestration"
+version = "0.0.1"
+description = "This is the dagster orchestration project for DAP pipelines"
+authors = ["Quazi Hoque <you@example.com>"]
```
feel free to change the email address to you broad email address

raaidbroad commented on orchestration/dap_orchestration/pipelines.py:
```diff
@@ -0,0 +1,45 @@
+    name="prod",
+    resource_defs={
+        "beam_runner": preconfigure_resource_for_mode(local_beam_runner, "prod", local_beam_runner_run_schema),
+        "refresh_directory": preconfigure_resource_for_mode(refresh_directory, "dev", refresh_dir_schema),
```
is this supposed to be dev?

raaidbroad commented on orchestration/dap_orchestration/pipelines.py:
```diff
@@ -0,0 +1,45 @@
+prod_mode = ModeDefinition(
+    name="prod",
+    resource_defs={
+        "beam_runner": preconfigure_resource_for_mode(local_beam_runner, "prod", local_beam_runner_run_schema),
```
Is this supposed to use the local beam runner?

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+)
+def extract_records(context: AbstractComputeExecutionContext) -> None:
+    """
+    :return: Returns the path to extracted files.
```
docstring needs updating, this returns `None`, not a path to extracted files?

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+    return _build_extract_config(
+        config,
+        "org.broadinstitute.monster.dap.hles.HLESurveyExtractionPipeline",
+        f"/extract/hles"
```
don't need the `f` before the string here

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+        "output_prefix": output_prefix,
+        "end_time": config["end_time"],
+        "api_token": config["api_token"],
+        "target_class": target_class,
```
don't need the trailling comma at the end of this line

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+)
+def transform_records(context: AbstractComputeExecutionContext) -> None:
+    """
+    :return: Returns the path to the transformation output json files.
```
docstring might need updating, it doesn't actually return a path

aherbst-broad reviewed (approved):
We paired on pushing this over the line, I think this lgtm now. Just need to get the formatting checks passing.

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+)
+def write_outfiles(context: AbstractComputeExecutionContext) -> None:
+    """
+    :return: Returns the path to the tsv outfiles.
```
doesn't return a path, docstring needs updating

quazi-h committed:
Removing unhelpful comments

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+    context.resources.beam_runner.run(arg_dict)
+
+
+def _build_transform_config(target_class: str, input_prefix: str, output_prefix: str) -> dict[str, str]:
```
I'm not sure how useful this function is as an abstraction; it simple takes the inputs and puts them in a dictionary, which we might as well just do in each of the configured functions below. Thoughts?

aherbst-broad replied:
@raaidbroad same pushback I had on `_build_extract_config`

raaidbroad reviewed (commented):
Here are some comments so far, adding more

raaidbroad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,160 @@
+            "experiments": "shuffle_mode=service",
+        }
+
+    def run(
```
nit: these args can comfortably fit on one line, so they should be formatted as such

raaidbroad commented on orchestration/dap_orchestration/resources.py:
```diff
@@ -0,0 +1,160 @@
+
+@resource
+def test_outfiles_writer(init_context: InitResourceContext) -> TestOutfilesWriter:
+    return TestOutfilesWriter()
```
add newline to EOF

raaidbroad commented on orchestration/dap_orchestration/solids.py:
```diff
@@ -0,0 +1,133 @@
+    context.resources.beam_runner.run(arg_dict)
+
+
+def _build_extract_config(config: dict[str, str], target_class: str, output_prefix: str) -> dict[str, str]:
```
I'm not sure how useful an abstraction this is since we're just building a dictionary; do you think it'd be more readable to just return a dictionary in each of the configured solids below?

aherbst-broad replied:
I'm going to push back on this; the premise was to extract the thing that was varying between the 3 flavors of solids (the target_class) and centralize the rest of the config management. I think this is worth keeping for now.

raaidbroad reviewed (commented):
Breaking up the resources into multiple files within a resources package would be more readable imo (same for solids but less important since those are atomic little nuggets anyway)

quazi-h committed:
Added config files for dataflow runner resource
Added a local mode to pipelines
Mypy, pep8 fixes

quazi-h committed:
"Added more accurate docstrings to the solids"

quazi-h committed:
Uncommented the environment solids being run in the pipelines script (testing)

quazi-h committed:
Pep8
